### ISSUE TITLE
# Spike-Timing Backpropagation

### ISSUE THREAD

**Issue by Zuzu3290** [Zuzu3290/SNNs-auf-GPUs#40]
## Objective
Replace or augment the current surrogate gradient (ATan) with a timing-based gradient — computing `dT_spike / dW` — so that the network learns from *when* neurons fire, not just *how often*.

## Description
The current training loop uses surrogate gradient BPTT: a smooth approximation of the spike threshold's derivative with respect to membrane potential. This is a rate-coding assumption — information lives in spike counts.

Spike-timing backpropagation treats the **firing time** as the signal. The gradient chain becomes:

```
dL/dW = dL/dT · dT/dW
```

where `T` is the time a neuron fires. This aligns with how neuromorphic hardware (Loihi, SpiNNaker) communicates internally — as timestamped events, not dense spike counts.

**Reference studies to review before implementing:**
- SpikeProp — Bohte et al. (2000) — first temporal backprop formulation
- EventProp — Kaiser et al. (2022) — exact adjoint-based gradients through spike times
- TTFS (Time-to-First-Spike) coding — spike time encodes class confidence directly

## Current State
| Component | Status |
|---|---|
| Surrogate gradient (ATan) | Active — all three frameworks |
| STDP regularizer | Active — soft timing constraint alongside BPTT |
| Timing gradient `dT/dW` | Not implemented |

The STDP term already rewards causal spike ordering. Timing backprop would make the gradient signal itself timing-based.

## Constraints to Resolve Before Implementation
1. **Silent neuron problem** — `dT/dW` is undefined when a neuron never fires. Activity regularization mitigates this but must be verified to hold under TRADES adversarial inputs.
2. **Framework support** — SNNTorch has the most extension points (`spike_grad` argument). Norse and SpikingJelly would need a custom autograd function.
3. **TRADES compatibility** — the adversarial perturbation loop expects differentiable logits; timing-coded output needs a compatible decode step.

### Phase 1 — Research
- [ ] Read EventProp (Kaiser 2022) and map `dT/dW` derivation to the LIF equations used here
- [ ] Identify whether SpikingJelly `IzhikevichNode` or SNNTorch `Alpha` exposes first-spike time
- [ ] Check if Norse `LIFCell` state carries enough information for timing gradients

### Phase 2 — Isolated prototype (non-TRADES path only)
- [ ] Write a custom `torch.autograd.Function` — `TimingGrad` — implementing `dT/dW`
- [ ] Plug into `snn_torch.py` as an alternative `spike_grad`, gated by a new config key `TIMING_GRAD_ENABLED`
- [ ] Validate on N-MNIST: confirm loss decreases and at least one spike fires per sample per forward pass

### Phase 3 — Integration
- [ ] Extend to Norse and SpikingJelly backends
- [ ] Verify compatibility with TRADES (replace or wrap `aggregate_spike_output` for timing decode)
- [ ] Benchmark: compare accuracy and spike rate vs surrogate gradient baseline

## Config Key (to add)
```yaml
timing_grad_enabled: false   # swap ATan surrogate for dT/dW timing gradient
```
